src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSourceHelper.OSX.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace System.Diagnostics.Tracing
+{
+    internal sealed class RuntimeEventSourceHelper
+    {
+        // from sys\resource.h
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct rusage_info_v3
+        {
+            internal fixed byte     ri_uuid[16];
+            internal ulong          ri_user_time;
+            internal ulong          ri_system_time;
+            internal ulong          ri_pkg_idle_wkups;
+            internal ulong          ri_interrupt_wkups;
+            internal ulong          ri_pageins;
+            internal ulong          ri_wired_size;
+            internal ulong          ri_resident_size;
+            internal ulong          ri_phys_footprint;
+            internal ulong          ri_proc_start_abstime;
+            internal ulong          ri_proc_exit_abstime;
+            internal ulong          ri_child_user_time;
+            internal ulong          ri_child_system_time;
+            internal ulong          ri_child_pkg_idle_wkups;
+            internal ulong          ri_child_interrupt_wkups;
+            internal ulong          ri_child_pageins;
+            internal ulong          ri_child_elapsed_abstime;
+            internal ulong          ri_diskio_bytesread;
+            internal ulong          ri_diskio_byteswritten;
+            internal ulong          ri_cpu_time_qos_default;
+            internal ulong          ri_cpu_time_qos_maintenance;
+            internal ulong          ri_cpu_time_qos_background;
+            internal ulong          ri_cpu_time_qos_utility;
+            internal ulong          ri_cpu_time_qos_legacy;
+            internal ulong          ri_cpu_time_qos_user_initiated;
+            internal ulong          ri_cpu_time_qos_user_interactive;
+            internal ulong          ri_billed_system_time;
+            internal ulong          ri_serviced_system_time;
+        }
+
+        // Constant from sys\resource.h
+        private const int RUSAGE_INFO_V3 = 3;
+        private const string OSX_LIBPROC = "libproc";
```

Comment:
We do not want to PInvoke the Unix APIs directly. They have unstable ABI.

We always PInvoke thin shims over the Unix ABI in `System.Native*` libraries. More details: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md#unix-shims

I think you should be able to use the existing `SystemNative_GetCpuUtilization` shim for this: https://github.com/dotnet/corefx/search?q=SystemNative_GetCpuUtilization&unscoped_q=SystemNative_GetCpuUtilization . Note that it is fine to use it in CoreLib even though it is implemented in CoreFX.

Reply:
Thanks for the suggestion Jan! And sorry for the delayed response :-( I was looking at adding some other APIs and came back to this recently. 

Quick question about that surface though -- It seems to me that this surface seems to be calculating "CPU usage in percentage". In particular, these lines (https://github.com/dotnet/corefx/blob/c1328f49f69d501618f1d867265efacedf107f23/src/Native/Unix/System.Native/pal_time.c#L194):
```
    if (cpuTotalTime > 0 && cpuBusyTime > 0)
    {
        cpuUtilization = (int32_t)(cpuBusyTime / cpuTotalTime);
    }
```
seems like it's calculating the ratio of busy time vs total time instead of the percentage. And since it's casted to an int, it always returns 0 when I tested it. 

Am I misinterpreting the intent of this surface? Or is this an actual issue?

Reply:
You can add a new API to the shim; or fixup the existing one to be more general - whatever works better.

My key point is that you should be able to just call the API, without parsing a bunch of stuff.

Reply:
Got it. Thanks!